File: CODE/DisplayDataframeWeb.py
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: Example_21_DisplayDataframeWeb.py
# DATE 2025-10-25
# DESCRIPTION: 

import streamlit as st
import pandas as pd
import sqlite3
import os

# Set page config
st.set_page_config(page_title="Dataframe Display", layout="wide")

# Connect to database and fetch data
@st.cache_data
def load_data():
    path_to_database = os.path.join(os.getenv("EXAMPLES", ""), "Boxscores.db")
    # Alternate path: path_to_database = "Boxscores.db"
    
    con = sqlite3.connect(path_to_database)
    query = "SELECT * FROM Boxscores"
    df = pd.read_sql_query(query, con)
    con.close()
    
    return df

# Load data
results_df = load_data()

# Title
st.title("Dataframe Display")

# Sidebar
with st.sidebar:
    st.header("Data Information")
    st.text(f"Rows: {len(results_df)} | Columns: {len(results_df.columns)}")
    st.divider()
    st.write("Use the search box and column filters to explore the data.")

# Main panel
st.subheader("Data Table")

# Add search functionality
search_term = st.text_input("Search across all columns:", "")

# Filter dataframe based on search
if search_term:
    mask = results_df.astype(str).apply(
        lambda x: x.str.contains(search_term, case=False, na=False)
        ).any(axis=1)

    filtered_df = results_df[mask]
else:
    filtered_df = results_df

# Display dataframe with built-in sorting and filtering
st.dataframe(
    filtered_df,
    hide_index=True
)

# Display row count after filtering
st.caption(f"Showing {len(filtered_df)} of {len(results_df)} rows")

################################################################################
# To run this, in terminal type: streamlit run /absolute path to this program/ #
################################################################################

File: CODE/EXTRA/DisplayDataframeWeb.R
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: DisplayDataframeWeb.R
# DATE 2025-10-25
# DESCRIPTION: 
# This program is a web application built with R Shiny, a framework used 
# to create interactive data dashboards. It allows users to view, search, 
# and filter the "Boxscores" database through a professional web interface.

library(shiny)
library(DT)
library(RSQLite)
library(DBI)
library(dplyr)

# Connect to database
path_to_database <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
con <- dbConnect(RSQLite::SQLite(), path_to_database)

query <- tbl(con, "Boxscores") |>
           select(everything())

results_df <- query |>
  collect()
  
dbDisconnect(con)

# UI
ui <- fluidPage(
  titlePanel("Dataframe Display"),
  
  sidebarLayout(
    sidebarPanel(
      h4("Data Information"),
      textOutput("info"),
      hr(),
      p("Use the search box and column filters to explore the data.")
    ),
    
    mainPanel(
      DTOutput("table")
    )
  )
)

# Server
server <- function(input, output, session) {
  
  output$info <- renderText({
    paste("Rows:", nrow(results_df), "| Columns:", ncol(results_df))
  })
  
  output$table <- renderDT({
    datatable(
      results_df,
      options = list(
        pageLength = 10,
        searching = TRUE,
        ordering = TRUE
      ),
      filter = "top",
      rownames = FALSE
    )
  })
}

# Run the app
shinyApp(ui = ui, server = server)

################################################################################################
# To run: In editor: (1) Highlight line 59 (2) Hit Ctrl-Enter                                  #
################################################################################################
